### ISSUE TITLE
Fused matmul untilize_out loses fp32 precision

### ISSUE THREAD

**Issue by fplavecTT** [tenstorrent/tt-metal#49836]
## Summary

When `ttnn.matmul` runs with `untilize_out=True` and `fp32_dest_acc_en=True` FP32 precision is not preserved. This becomes apparent with input intentionally crafted so that exact result relies on cancellation across the K reduction; the output has PCC ~ 0.13, even though the same matmul without `untilize_out` is accurate (PCC ~ 0.996). The loss is attributable to the `untilize_out` output path, independent of how the K reduction is accumulated.

## Reproduction
`tests/ttnn/unit_tests/operations/matmul/test_matmul.py::test_matmul_fp32_crossblock_reload_untilize_precision` (currently marked `xfail`).

## Observed (Blackhole)

| Accumulation path | `untilize_out` | PCC |
|---|---|---|
| reload (`packer_l1_acc=False`) | `True` | ~ 0.12 |
| packer L1 (`packer_l1_acc=True`, no reload) | `True` | ~ 0.13 |
| either path (`packer_l1_acc =` `False` or `True`) | `False` | ~ 0.996 (passes) |

## What this isolates

`packer_l1_acc=True` performs the K accumulation in fp32 in SRAM with no cross-block reload, yet still yields PCC ~ 0.13 with `untilize_out=True`. Since both accumulation methods land at low PCC and the non-`untilize_out` variant passes, the accuracy loss originates in the `untilize_out` output path (the fused reblock/untilize stage), not in the K-reduction accumulation.

**Comment by iwroszTT** [comment]
## This is not a precision bug. It is a layout bug.

I investigated this while checking whether it shared a root cause with another matmul accumulation bug. It does not. The title is misleading and will send people down the wrong path, as it did me.

**No data is lost. The values are correct but in the wrong positions.**

Comparing the `untilize_out=True` output against the `untilize_out=False` output of the same matmul:

- `sorted(bad)` vs `sorted(good)`: max absolute difference **0.000000**, exact multiset match.
- At 16-element granularity: an exact 1:1 permutation of all 256 chunks, none unmapped.
- First differing element is at flat index **16**, which is the face width.

A precision loss cannot produce a bit-exact multiset. This is a permutation.

## The bug is much wider than this test shows

The test's setup — offset cancellation, K=8192, fp32 accumulation — is all incidental. The failure does not depend on any of it:

| varied | result |
|---|---|
| `num_blocks` = 8, 32, 128 | identical to 5 significant figures |
| `packer_l1_acc` = False, True | identical |
| `fp32_dest_acc_en` = False, True | identical |
| every block geometry tried | all fail, down to a single 32x32 tile (`per_core_M = per_core_N = 1`, PCC 0.027) |

Minimal repro, plain random inputs, no fp32, K=256, one tile:

| | untilize off | untilize on |
|---|---|---|
| fp32 off | PCC 0.99997 | **PCC 0.036** |
| fp32 on | PCC 0.99999 | **PCC 0.036** |

So `untilize_out` is broken for this config in general, not only under precision stress.

## What I ruled out

Three candidate causes, all tested and all wrong:

1. **Intermediate buffer sized wrong.** No. The 1D factory sets `interm0_CB_tiles = out_block_tiles` and keeps `out_CB_tiles` separate. This is correct.
2. **Output declared TILE while holding row-major data.** No. The output tensor already reports `Layout.ROW_MAJOR`. I changed the interleaved fall-through in `compute_output_specs` anyway; it produced a *different* wrong answer (PCC 0.135 -> -0.026), so that line is not the defect.
3. **`PageConfig(ROW_MAJOR, output_tile)` carrying an inconsistent tile spec.** No. Dropping the tile when `untilize_out`, matching what the sharded branch already does, was a no-op: PCC returned to exactly 0.134595.

## Where I think the problem is

There is a three-way mismatch on the interleaved output path:

- The compute kernel untilizes the block, so the **content is row-major**.
- The writer (`reader_bmm_tile_layout_in1_sender_writer_padding.cpp`) addresses the output **by tile id** — `out_tensor_start_tile_id`, `out_tensor_stride_w` / `_h`, `TensorAccessor` over the output — so it writes **tile-sized pages**.
- The tensor spec says **ROW_MAJOR**.

These three cannot all be right. My host-side attempts kept moving the corruption around rather than removing it, which fits: the fix likely needs the writer to use row-major stick addressing when `untilize_out` is set, not a change in the spec.

## Two suggestions

1. **Retitle the issue.** Something like "Fused matmul untilize_out produces wrong output layout". The current title costs anyone who picks this up a detour through accumulation and dest-register precision.
2. **Check whether `untilize_out` works on a sharded output path.** I did not test this — my attempt hit an unrelated `TT_FATAL` (HEIGHT_SHARDED rejected on a 1-row grid). If sharded works and interleaved does not, then the cheap fix is to reject the interleaved combination in `validate_matmul_untilize_out`, which currently allows it.

The existing xfailed test is a poor reproducer for a layout bug, because its numerics machinery hides the fact that the values are bit-exact. A five-line test with random inputs and one tile shows it more clearly.

**Comment by iwroszTT** [comment]
## Reproducer

Drop this in `tests/ttnn/unit_tests/operations/matmul/`. It runs in under 2 seconds on one core and needs no special setup.

It has two tests. The first is the regression test. The second shows *why* it fails.

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""Reproducer for #49836: fused matmul untilize_out returns the right values in the wrong places."""

import pytest
import torch
import ttnn

from models.common.utility_functions import comp_pcc


def _matmul_one_tile(device, untilize_out):
    """Smallest matmul that selects the 1D multicast factory: one core, one 32x32 output tile.

    Returns (torch_reference, device_result).
    """
    m, k, n = 32, 256, 32
    torch.manual_seed(0)
    a = torch.randn(m, k).bfloat16()
    b = torch.randn(k, n).bfloat16()

    program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
        compute_with_storage_grid_size=ttnn.CoreCoord(1, 1),
        in0_block_w=2,
        out_subblock_h=1,
        out_subblock_w=1,
        per_core_M=1,
        per_core_N=1,
        fuse_batch=True,
        mcast_in0=True,
        untilize_out=untilize_out,
    )
    compute_kernel_config = ttnn.WormholeComputeKernelConfig(
        math_fidelity=ttnn.MathFidelity.HiFi4,
        math_approx_mode=False,
        fp32_dest_acc_en=False,
        packer_l1_acc=False,
    )
    a_t = ttnn.from_torch(
        a, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    b_t = ttnn.from_torch(
        b, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16, memory_config=ttnn.DRAM_MEMORY_CONFIG
    )
    out = ttnn.matmul(
        a_t,
        b_t,
        program_config=program_config,
        compute_kernel_config=compute_kernel_config,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
        dtype=ttnn.bfloat16,
    )
    return a.float() @ b.float(), ttnn.to_torch(out).float().reshape(m, n)


@pytest.mark.xfail(reason="untilize_out scrambles the output layout. Issue #49836", strict=True)
def test_matmul_untilize_out_matches_reference(device):
    """The regression test. Remove the xfail when #49836 is fixed."""
    ref, got = _matmul_one_tile(device, untilize_out=True)
    passed, message = comp_pcc(ref, got, 0.99)
    assert passed, message


def test_matmul_untilize_out_is_a_permutation(device):
    """Diagnostic: shows the failure is a layout scramble, not a precision loss.

    Compares untilize_out=True against untilize_out=False on identical inputs. Every value is
    bit-identical between the two; only the positions differ. This test documents the nature of
    #49836 and should be deleted along with the xfail above once it is fixed.
    """
    _ref, good = _matmul_one_tile(device, untilize_out=False)
    _ref2, bad = _matmul_one_tile(device, untilize_out=True)

    good_sorted = torch.sort(good.flatten()).values
    bad_sorted = torch.sort(bad.flatten()).values
    max_diff = (good_sorted - bad_sorted).abs().max().item()

    first_diff = torch.nonzero((good.flatten() - bad.flatten()).abs() > 0).flatten()
    print(f"\nuntilize_out=False vs True, identical inputs:")
    print(f"  PCC between them            : {comp_pcc(good, bad, 0.99)[1]}")
    print(f"  max diff of sorted values   : {max_diff}")
    print(f"  same multiset of values     : {bool(torch.equal(good_sorted, bad_sorted))}")
    print(f"  first differing flat index  : {first_diff[0].item() if first_diff.numel() else 'none'}")
    print(f"  elements out of place       : {first_diff.numel()} of {good.numel()}")

    # Nothing is lost: the two runs hold exactly the same numbers.
    assert torch.equal(good_sorted, bad_sorted), "values differ, so this is not purely a layout bug"
```

## Output on Wormhole today

```
untilize_out=False vs True, identical inputs:
  PCC between them            : 0.02711300122248392
  max diff of sorted values   : 0.0
  same multiset of values     : True
  first differing flat index  : 16
  elements out of place       : 960 of 1024

1 passed, 1 xfailed in 1.68s
```

Read it as:

- `max diff of sorted values = 0.0` and `same multiset = True`. The two runs hold exactly the same 1024 numbers. Nothing is lost, so this is not a precision problem.
- `960 of 1024` elements are in the wrong place, and `PCC = 0.027`.
- The first 16 elements are correct and the 17th is not. 16 is the face width, which is where the layout handling goes wrong.

## Notes

- No fp32, no offset cancellation, no deep K. One output tile on one core is enough. The existing `test_matmul_fp32_crossblock_reload_untilize_precision` needs none of its setup to show the bug, and its numerics machinery hides that the values are bit-exact.
- The xfail is `strict=True`, so it turns into a failure the moment the bug is fixed. That is the signal to delete both tests.
- The second test passes today and asserts only the property that makes this a layout bug. If a future change makes the values differ as well, it fails and tells you the problem has changed shape.
